### DISCUSSION TITLE
Containerfile versus Dockerfile, What's the difference?

### DISCUSSION THREAD

**Discussion by TomSweeneyRedHat** [containers/buildah#3170] [General]
As I write this, it's April 20, 2021 and the latest release of Buildah stands at v1.20.1. 

We've been asked a couple of times over the past few weeks, "What's the difference between a Containerfile and a Dockefile?"  First off, as you can tell, they have different names.  Secondly, that's it, well at least in theory.  Read on for more details.

The content, commands, and any features that are usable in a Dockerfile should also be usable in a Containerfile too.  That's the design goal.  There may be newer options or commands in a Dockerfile that are not usable in a Containerfile.  If you find any that you need that are missing, please log an [issue](https://github.com/containers/buildah/issues) and hopefully we can add them to the Containerfile processing code too.

The one small functional difference between the two files is how Buildah and Podman use the files.  If there's a `Containerfile` in the context directory, Podman and Buildah will use that for the build.  If there's a `Dockerfile` in the context directory, the build will use that if there is not a  `Containerfile` present.  If both a `Containerfile` and a `Dockerfile` are in the context directory, Podman and Buildah will use the `Containerfile` and will ignore the `Dockerfile`.  Then if you use the `--file` option, you can name the file pretty much anything you want.

Given all that, at some point in time in the future if we find commands that would be helpful in a Containerfile that are not present in a Dockerfile, then if the community desires, they might be added to the Containerfile processing.   There are currently no plans for that and (KOD), I don't anticipate changes like that in the foreseeable future.

Hope that helps to explain the differences between a Containerfile and a Dockefile for Podman and Buildah.

**Comment by TomSweeneyRedHat** [comment]
@isavl thanks for the poke via your recent issue, a little more detail on Containerfile vs Dockerfile if you're interested.

**Comment by rhatdan** [comment]
LGTM

**Comment by j3mdamas** [comment]
Hi,
I have been doing some research, but maybe you can point me in the right direction.
`buildah` builds OCI-compliant images. There are good references for it: https://github.com/opencontainers/image-spec/blob/main/spec.md. But when trying to have a good reference for Containerfile, I always end up using [this](https://docs.docker.com/engine/reference/builder/). Is that the right reference for the file format? Is there some OCI-ish reference? I could not find it.

**Comment by TomSweeneyRedHat** [comment]
@j3mdamas, the innards of a Containerfile and a Dockerfile should be equivalent, so the Dockerfile Reference page that you pointed to works.  The only other document that we have is the Containerfile [manpage](https://github.com/containers/common/blob/main/docs/Containerfile.5.md), but that's admittedly not as detailed.

**Reply by j3mdamas** [reply]
I see, thanks for the feedback :)

**Comment by imphil** [comment]
With rather sparse documentation on what syntax buildah vs. buildkit support, I have a similar question: How does buildah deal with [external frontends](https://hub.docker.com/r/docker/dockerfile), specified through `# syntax=docker/dockerfile:1.2`? Does it ignore this statement? And if so, what version of the Dockerfile syntax does buildah support?

**Reply by rhatdan** [reply]
Buildah is attempting to follow buildkit syntax.  We ignore the syntax flag.

**Reply by imphil** [reply]
Thanks! Is there documentation somewhere what version of the syntax a particular buildah version supports?

**Reply by rhatdan** [reply]
No we don't track it that closely, although I would guess you could approximately line up the support that buildkit had. 
The bottom line is we add features to Buildah when people realize that a Dockerfile works on Docker but not on buildah and then we add support, if their is a difference between old Docker and Buildkit, then we go with buildkits version.
